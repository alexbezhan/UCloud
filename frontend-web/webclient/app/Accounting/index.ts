import Usage from "Accounting/Usage";

export {Usage};