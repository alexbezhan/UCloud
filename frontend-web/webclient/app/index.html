<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
