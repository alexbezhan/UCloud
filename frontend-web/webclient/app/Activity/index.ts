import { ActivityReduxObject } from "DefaultObjects";
import { ScrollRequest } from "Scroll";

export enum ActivityType {
    DOWNLOAD = "download",
    UPDATED = "updated",
    DELETED = "deleted",
    FAVORITE = "favorite",
    INSPECTED = "inspected",
    MOVED = "moved"
}

export interface ActivityGroup {
    type: ActivityType;
    newestTimestamp: number;
    numberOfHiddenResults: number | null;
    items: Activity[];
}

export interface Activity {
    type: ActivityType;
    timestamp: number;
    fileId: string;
    username: string;
    originalFilePath: string;
}

export interface ActivityFilter {
    collapseAt?: number;
    type?: ActivityType;
    minTimestamp?: Date;
    maxTimestamp?: Date;
}

export interface FavoriteActivity extends Activity {
    favorite: boolean;
}

export interface MovedActivity extends Activity {
    newName: string;
}


export interface ActivityDispatchProps {
    onMount: () => void;
    fetchActivity: (scroll: ScrollRequest<number>, filter?: ActivityFilter) => void;
    resetActivity: () => void;
    setRefresh: (refresh?: () => void) => void;
    updateFilter: (filter: Partial<ActivityFilter>) => void;
}

export interface ActivityOwnProps {
}

export type ActivityProps = ActivityReduxObject & ActivityDispatchProps & ActivityOwnProps;
