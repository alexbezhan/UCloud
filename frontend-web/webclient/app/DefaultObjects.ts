import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState, SortOrder} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.DESCENDING
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
