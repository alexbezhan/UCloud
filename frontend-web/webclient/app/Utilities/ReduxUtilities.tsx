import {createStore, combineReducers, Store, AnyAction, Action} from "redux";
import {composeWithDevTools} from 'redux-devtools-extension';
import {ReduxObject, initObject} from "DefaultObjects";
import {responsiveBP} from "ui-components/theme";
import {createResponsiveStateReducer} from "redux-responsive";
import {Cloud} from "Authentication/SDUCloudObject";
import {CONTEXT_SWITCH, USER_LOGOUT, USER_LOGIN} from "Navigation/Redux/HeaderReducer";

export function configureStore(initialObject: Partial<ReduxObject>, reducers, enhancers?): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject(Cloud.homeFolder);
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);