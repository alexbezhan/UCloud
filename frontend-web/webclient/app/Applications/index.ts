import {AnalysisReduxObject, ResponsiveReduxObject} from "DefaultObjects";
import {SortOrder} from "Files";
import {History} from "history";
import {SetStatusLoading} from "Navigation/Redux/StatusActions";
import PromiseKeeper from "PromiseKeeper";
import * as React from "react";
import {SharedFileSystemMount} from "Applications/FileSystems";
import {match} from "react-router";
import {ParameterValues} from "Utilities/ApplicationUtilities";

export interface Analysis {
    name: string;
    checked?: boolean;
    status: string;
    state: AppState;
    jobId: string;
    appName: string;
    appVersion: string;
    createdAt: number;
    modifiedAt: number;
    expiresAt?: number;
    owner: string;
    metadata: ApplicationMetadata
}

export type AnalysesStateProps = AnalysisReduxObject & {responsive: ResponsiveReduxObject}
export type AnalysesProps = AnalysesStateProps & AnalysesOperations;

type FetchJobsOperation = (
    itemsPerPage: number,
    pageNumber: number,
    sortOrder: SortOrder,
    sortBy: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: AppState
) => void;

export interface AnalysesOperations {
    setLoading: (loading: boolean) => void;
    fetchJobs: FetchJobsOperation;
    onInit: () => void;
    setRefresh: (refresh?: () => void) => void;
    checkAnalysis: (jobId: string, checked: boolean) => void;
    checkAllAnalyses: (checked: boolean) => void;
}

export interface DetailedResultOperations {
    setPageTitle: (jobId: string) => void;
    setLoading: (loading: boolean) => void;
    setRefresh: (refresh?: () => void) => void;
}

export interface DetailedResultProps extends DetailedResultOperations {
    match: match<{jobId: string}>;
    history: History;
}

export interface Application {
    favorite: boolean;
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: ApplicationDescription;
    tool: ApplicationTool;
    imageUrl: string;
}

interface ApplicationTool {
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: {
        info: ApplicationInfo
        container: string
        defaultNumberOfNodes: number
        defaultTasksPerNode: number
        defaultAllocationTime: MaxTime
        requiredModules: string[]
        authors: string[]
        title: string
        description: string
        backend: string
        license: string
    };
}

interface ApplicationInfo {
    name: string;
    version: string;
}

export interface ApplicationDescription {
    info: ApplicationInfo;
    tool: ApplicationInfo;
    authors: string[];
    title: string;
    description: string;
    invocation: any[]; // FIXME: Add type
    parameters: ApplicationParameter[];
    outputFileGlobs: string[];
    website?: string;
    resources: {multiNodeSupport: boolean};
    tags: string[];
}

export enum AppState {
    VALIDATED = "VALIDATED",
    PREPARED = "PREPARED",
    SCHEDULED = "SCHEDULED",
    RUNNING = "RUNNING",
    TRANSFER_SUCCESS = "TRANSFER_SUCCESS",
    SUCCESS = "SUCCESS",
    FAILURE = "FAILURE",
    CANCELLING = "CANCELLING"
}

export interface DetailedResultState {
    name: string;
    complete: boolean;
    appState: AppState;
    failedState?: AppState;
    status: string;
    app?: ApplicationMetadata;
    stdout: string;
    stderr: string;
    stdoutLine: number;
    stderrLine: number;
    reloadIntervalId: number;
    promises: PromiseKeeper;
    outputFolder?: string;
    appType?: ApplicationType;
    webLink?: string;
    timeLeft: number | null;
}

export type StdElement = {scrollTop: number, scrollHeight: number} | null

export interface MaxTime {
    hours: number;
    minutes: number;
    seconds: number;
}

export interface MaxTimeForInput {
    hours: number;
    minutes: number;
    seconds: number;
}

export interface JobSchedulingOptionsForInput {
    maxTime: MaxTimeForInput;
    numberOfNodes: number;
    tasksPerNode: number;
    name: React.RefObject<HTMLInputElement>;
}

export interface AdditionalMountedFolder {
    readOnly: boolean;
    ref: React.RefObject<HTMLInputElement>;
    defaultValue?: string;
}

export interface AdditionalPeer {
    nameRef: React.RefObject<HTMLInputElement>;
    jobIdRef: React.RefObject<HTMLInputElement>;
}

export interface RunAppState {
    promises: PromiseKeeper;
    jobSubmitted: boolean;
    initialSubmit: boolean;
    application?: FullAppInfo;
    parameterValues: ParameterValues;
    schedulingOptions: JobSchedulingOptionsForInput;
    favorite: boolean;
    favoriteLoading: boolean;
    mountedFolders: AdditionalMountedFolder[];
    additionalPeers: AdditionalPeer[];
    fsShown: boolean;
    sharedFileSystems: { mounts: SharedFileSystemMount[] };
}

export interface RunOperations extends SetStatusLoading {
    updatePageTitle: () => void;
}

export interface RunAppProps extends RunOperations {
    match: match<{appName: string, appVersion: string}>;
    history: History;
    updatePageTitle: () => void;
}

export interface NumberParameter extends BaseParameter {
    defaultValue: {value: number, type: "double" | "int"} | null;
    min: number | null;
    max: number | null;
    step: number | null;
    type: ParameterTypes.Integer | ParameterTypes.FloatingPoint;
}

export interface BooleanParameter extends BaseParameter {
    defaultValue: {value: boolean, type: "bool"} | null;
    trueValue?: string | null;
    falseValue?: string | null;
    type: ParameterTypes.Boolean;
}

export interface InputFileParameter extends BaseParameter {
    defaultValue: string | null;
    type: ParameterTypes.InputFile;
}

export interface InputDirectoryParameter extends BaseParameter {
    defaultValue: string | null;
    type: ParameterTypes.InputDirectory;
}

export interface TextParameter extends BaseParameter {
    defaultValue: {value: string, type: "string"} | null;
    type: ParameterTypes.Text;
}

export interface PeerParameter extends BaseParameter {
    suggestedApplication: string | null
    type: ParameterTypes.Peer
}

export interface SharedFileSystemParameter extends BaseParameter {
    fsType: "EPHEMERAL" | "PERSISTENT"
    mountLocation: string
    exportToPeers: boolean
    type: ParameterTypes.SharedFileSystem
}

interface BaseParameter {
    name: string;
    optional: boolean;
    title: string;
    description: string;
    unitName?: string | React.ReactNode | null;
    type: string;
    visible?: boolean;
}

export type ApplicationParameter = 
	InputFileParameter | 
	InputDirectoryParameter | 
	NumberParameter | 
	BooleanParameter |
    TextParameter | 
    PeerParameter | 
    SharedFileSystemParameter;

type Invocation = WordInvocation | VarInvocation;

interface WordInvocation {
    type: "word";
    word: string;
}

interface VarInvocation {
    type: "var";
    variableNames: string[];
    prefixGlobal: string;
    suffixGlobal: string;
    prefixVariable: string;
    suffixVariable: string;
    variableSeparator: string;
}

interface Info {name: string; version: string;}
export interface Description {
    info: Info;
    tool: Info;
    authors: string[];
    title: string;
    description: string;
    invocation: Invocation[];
    parameters: ApplicationParameter[];
    outputFileGlobs: [string, string];
    tags: string[];
}

export enum ParameterTypes {
    InputFile = "input_file",
    InputDirectory = "input_directory",
    Integer = "integer",
    FloatingPoint = "floating_point",
    Text = "text",
    Boolean = "boolean",
    Peer = "peer",
    SharedFileSystem = "shared_file_system"
}

export interface SearchFieldProps {
    onSubmit: () => void;
    icon: string;
    placeholder: string;
    value: string;
    loading: boolean;
    onValueChange: (value: string) => void;
}

export interface DetailedApplicationSearchReduxState {
    hidden: boolean;
    appName: string;
    appVersion: string;
    tags: string;
    error?: string;
    loading: boolean;
}

export interface DetailedApplicationOperations {
    setAppName: (n: string) => void;
    setVersionName: (v: string) => void;
    fetchApplicationsFromName: (q: string, i: number, p: number, c?: Function) => void;
    fetchApplicationsFromTag: (t: string, i: number, p: number, c?: Function) => void;
}



// New interfaces
export interface ApplicationMetadata {
    name: string;
    version: string;
    authors: string[];
    title: string;
    description: string;
    tags: string[];
    website?: string;
}

type ApplicationType = "BATCH" | "VNC" | "WEB"

export interface ApplicationInvocationDescription {
    tool: Tool;
    invocation: Invocation[];
    parameters: ApplicationParameter[];
    outputFileGlobs: string[];
    applicationType: ApplicationType;
    shouldAllowAdditionalMounts: boolean
    shouldAllowAdditionalPeers: boolean
    allowMultiNode: boolean
}

interface Resources {
    multiNodeSupport: boolean;
    coreRequirements: number;
    memoryRequirementsMb: number;
    gpuRequirements: number;
    tempStorageRequirementsGb: number;
    persistentStorageRequirementsGb: number;
}

interface Tool {
    name: string;
    version: string;
    tool: ToolReference;
}

interface ToolReference {
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: ToolDescription;
}

interface NameAndVersion {
    name: string;
    version: string;
}

interface ToolDescription {
    info: NameAndVersion;
    container: string;
    defaultNumberOfNodes: number;
    defaultTasksPerNode: number;
    defaultTimeAllocation: MaxTime;
    requiredModules: string[];
    authors: string[];
    title: string;
    description: string;
    backend: string;
    license: string;
}

export interface WithAppMetadata {
    metadata: ApplicationMetadata;
}

export interface WithAppInvocation {
    invocation: ApplicationInvocationDescription;
}

export interface WithAppFavorite {
    favorite: boolean;
}

export enum RunsSortBy {
    state = "STATE",
    application = "APPLICATION",
    startedAt = "STARTED_AT",
    lastUpdate = "LAST_UPDATE",
    createdAt = "CREATED_AT",
    name = "NAME"
}
export interface WithAllAppTags {
    tags: string[];
}

export type FullAppInfo = WithAppFavorite & WithAppInvocation & WithAppMetadata & WithAllAppTags;
