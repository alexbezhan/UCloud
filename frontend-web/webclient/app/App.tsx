import * as React from "react";
import * as ReactDOM from "react-dom";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import fileInfo from "Files/Redux/FileInfoReducer";
import {theme, UIGlobalStyle} from "ui-components";
import {createGlobalStyle, ThemeProvider} from "styled-components";
import {Cloud} from "Authentication/SDUCloudObject";
import {initObject} from "DefaultObjects";
import Core from "Core";
import avatar from "UserSettings/Redux/AvataaarReducer";
import header, {USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH} from "Navigation/Redux/HeaderReducer";
import status from "Navigation/Redux/StatusReducer";
import applications from "Applications/Redux/BrowseReducer";
import dashboard from "Dashboard/Redux/DashboardReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import analyses from "Applications/Redux/AnalysesReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import uploader from "Uploader/Redux/UploaderReducer";
import activity from "Activity/Redux/ActivityReducer";
import simpleSearch from "Search/Redux/SearchReducer";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import filePreview from "Files/Redux/FilePreviewReducer";
import * as AppRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import {configureStore} from "Utilities/ReduxUtilities";
import {responsiveStoreEnhancer, createResponsiveStateReducer} from 'redux-responsive';
import {responsiveBP, invertedColors} from "ui-components/theme";
import {findAvatar} from "UserSettings/Redux/AvataaarActions";
import Header from "Navigation/Header";
import {isLightThemeStored, setSiteTheme} from "UtilityFunctions";
import * as ProjectRedux from "Project/Redux";

const store = configureStore(initObject(Cloud.homeFolder), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    filePreview,
    ...AppRedux.reducers,
    ...AccountingRedux.reducers,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}

export function dispatchUserAction(type: typeof USER_LOGIN | typeof USER_LOGOUT | typeof CONTEXT_SWITCH) {
    store.dispatch({type})
}

export async function onLogin() {
    const action = await findAvatar();
    if (action !== null) store.dispatch(action);
}

const GlobalStyle = createGlobalStyle`
  ${() => UIGlobalStyle}
`;

Cloud.initializeStore(store);

function App({children}) {
    const [isLightTheme, setTheme] = React.useState(isLightThemeStored());
    const setAndStoreTheme = (isLight: boolean) => (setSiteTheme(isLight), setTheme(isLight));
    return (
        <ThemeProvider theme={isLightTheme ? theme : {...theme, colors: invertedColors}}>
            <>
                <GlobalStyle/>
                <BrowserRouter basename="app">
                    <Header toggleTheme={() => isLightTheme ? setAndStoreTheme(false) : setAndStoreTheme(true)}/>
                    {children}
                </BrowserRouter>
            </>
        </ThemeProvider>
    )
}

ReactDOM.render(
    <Provider store={store}>
        <App>
            <Core/>
        </App>
    </Provider>,
    document.getElementById("app")
);
