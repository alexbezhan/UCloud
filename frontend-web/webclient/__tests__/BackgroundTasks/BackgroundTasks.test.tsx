import * as React from "react";
import { create } from "react-test-renderer";
import BackgroundTasks from "../../app/BackgroundTasks/BackgroundTask";
import { Provider } from "react-redux";
import { configureStore } from "../../app/Utilities/ReduxUtilities";
import uploader from "../../app/Uploader/Redux/UploaderReducer";
import { initUploads } from "../../app/DefaultObjects";

describe("BackgroundTasks", () => {
    test("Mount", () => {

        expect(create(
            <Provider store={configureStore({ uploader: initUploads() }, { uploader })}>
                <BackgroundTasks />
            </Provider>)).toMatchSnapshot();
    });
});
