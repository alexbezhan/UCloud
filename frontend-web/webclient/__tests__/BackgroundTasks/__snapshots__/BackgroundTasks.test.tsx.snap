// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundTasks Mount 1`] = `null`;
